### DISCUSSION TITLE
Switch to GitHub Actions for CI

### DISCUSSION THREAD

**Discussion by QarthO** [GriefPrevention/GriefPrevention#2210] [Ideas]
In reference to this: https://github.com/GriefPrevention/GriefPrevention/issues/2190#issuecomment-1917831347

Feel free to use this [actions script](https://github.com/QarthO/Ancient-Mystery/blob/main/.github/workflows/maven.yml) I made for some of my plugins to build and upload the jar on every commit. It's drop and place with no setup needed for GP. And it should work with any plugin as long as it's using Maven.

However, you will need to add any test cases, you can also add the Maven cache plugin.
You can also modify the ``on:`` section at the top to customize when and what branches it'll build the jar as you please. 

It will also rename the jar to: ``<plugin>-<version>-<commit-sha>.jar`` (example: ``GriefPrevention-16.18.2-1ba7df4.jar``) But that's all it does outside of building the jar.

**Comment by RoboMWM** [comment]
So my main question with Actions is, what permissions does it have and can I control it?

I.e. I'd imagine it can work from any branch(?), and people can PR changes to it...

**Reply by QarthO** [reply]
Actions are only applied when things get pushed to the branch. (atleast this script)

So how you control is the same way you control PRs. This action script won't ever be called unless it's merged. And then its only ran everytime the ``on:`` case is met. So for my specific action script, it'll only be called everytime something is pushed to the ``main`` branch.
So for GP, would need to modify that to the ``master`` branch, or whatever branches you want this script to be ran on.

**Reply by QarthO** [reply]
Basically no one should be able to run an actions script, or even run this script, unless they're able to push to the branches specific in the ``on:`` case

**Reply by RoboMWM** [reply]
Ok, so it checks the on: case. 

What about permissions then? Do actions just always have administrative privileges to the repo they're run on?

**Reply by Jikoo** [reply]
Permissions are dependent on the triggering workflow. Most do have administrative permissions. In the case of PRs trying to modify actions, the original action is run and not the modified copy, if that's what you're worried about. `pull_request` has read-only permissions; `pull_request_target` has full permissions.

> same reason why I didn't want Dependabot on this repo

Dependabot only has permission to scrape your dependencies and submit PRs. I generally also automate accepting PRs that build with Actions, but I will admit that due to misunderstanding the difference between `pull_request_target` and `pull_request` I had it pull anything as long as the master branch compiled in my first few efforts. 🤦 Either way, Dependabot on its own can at worst only annoy you by telling you a lot of dependencies are out of date.

**Reply by RoboMWM** [reply]
I meant the old Dependabot that used to be able to merge PRs automatically, that has been changed now yes.

**Reply by RoboMWM** [reply]
Hmm, is it possible to restrict all actions to be read-only then? Say I accidentally pull in a PR that has a change to the workflow... I wouldn't want it to be able to now suddenly modify other things in the repo. 

As of right now, someone can change how appveyor builds are built in a PR (upon me pulling it), but no matter what they do, they can't make appveyor modify anything on github.

**Reply by QarthO** [reply]
I mean this would be an issue of you accidentally pulling the PR. This exact question can be used before the idea of actions. What happens if right now you accidentally pull a PR that puts malicious code into the source files?

Just like if someone does a bad PR for the source, it can be reverted.

**Reply by Jikoo** [reply]
Read-only is an option, yeah. https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#setting-the-permissions-of-the-github_token-for-your-repository

**Reply by QarthO** [reply]
This says by default its set to read-only

**Reply by Jikoo** [reply]
I do also kind of agree, if you're not pulling bad code you don't need to worry about that. If you're worried about missing a modification to actions files, we could actually set up Actions to confirm that no changes to Actions files are made before running. That would mean the required builds wouldn't complete, so you'd have to override to pull assuming the branch protection rules are set to require CI success. Additionally, in the case of Actions not having restricted permissions, it could comment to alert you that changes were made.
/e: actually, not sure that Actions needs write perms to make comments. Would have to test that more.

**Reply by QarthO** [reply]
Also, looks like you can prevent actions from creating and approving PRs, which we should 100% make sure is done (which by default prevents)

**Reply by QarthO** [reply]
Overall looking at the docs u sent, seems like all the defaults for actions are setup in a way to prevent any malicious thing from happening to your codebase

**Reply by RoboMWM** [reply]
> What happens if right now you accidentally pull a PR that puts malicious code into the source files?

The most that can happen is appveyor goes rogue, that's about it. Nothing else gets executed as it's just a webhook trigger, which that configuration is done outside of the repo itself (in the settings instead).

Mk, just took another look at the Actions configuration. Seems the default is/was read and write. Changed that now. Though, it only says the "default" permissions, and that granular permissions can be specified... in the YAML of the workflow. Why. ~~I guess I can disable write at the org level... would be nice if I could do at repo level too, but whatevs.~~ nevermind, this information is also conflicting... that anyone with write can just override this right in the workflow itself...??

**Reply by RoboMWM** [reply]
https://github.blog/2023-11-16-security-best-practices-for-authors-of-github-actions/

Well, nice.

Anyways, my primary reason isn't necessarily malicious code, but mistakes, such as accidentally editing/deleting Issues, PRs, etc. Doesn't seem the token will grant any access to repo settings, and I can't tell if it will prohibit any code pushes... just guessing at what can be granted by the names of the listed scopes.

**Reply by RoboMWM** [reply]
https://github.com/ossf/scorecard/blob/main/docs/checks.md#token-permissions

So no way to enforce it other than just ensure nobody touches that in a PR I guess. Guh. Seems maybe I could have workflows in a different repo to reduce chances of me pulling a PR that modifies Actions.

At the end of the day, is this just a matter of not liking AppVeyor or something and preferring Github's UI or whatever?

**Reply by nouish** [reply]
You are overthinking this. If typical use was dangerous, you wouldn't be seeing every repo on GitHub using it.

If your concern is pulling in a dangerous PR, that implies a _much_ bigger problem.

**Reply by Jikoo** [reply]
> At the end of the day, is this just a matter of not liking AppVeyor or something and preferring Github's UI or whatever?

Mostly the UI. Since GP releases are relatively stagnant, users turned to dev builds for fixes they considered critical, particularly when new MC versions released. AppVeyor does label which branch provides a build, but it does not have branch filtering as far as I can tell. Users were downloading the "latest" build which happened to be whatever was the most recent PR, some of which got declined. That potentially being a malicious pull is kinda concerning to me. Hasn't happened, might not ever happen, but you can imagine the fiasco that a dev build from the "official" source having a forceop would be.

Imo because of that limitation, AppVeyor should only be officially-sanctioned builds. That's where Actions can fill in the gap. If users want a specific branch they should download the resulting file from its most recent Actions run. Actions can be filtered based on branch, workflow, etc. The download process is also a little more annoying, which adds an extra barrier for users who know just enough to be a danger to themselves.

**Reply by RoboMWM** [reply]
Cool thanks for that input, I'll spawn new comment threads.

**Reply by RoboMWM** [reply]
> AppVeyor does label which branch provides a build, but it does not have branch filtering as far as I can tell.

Do you mean filtering as in which branches publish an artifact?

**Reply by Jikoo** [reply]
> Do you mean filtering as in which branches publish an artifact?

The user view can't be filtered afaik - no way to provide a link that will only display builds from the master branch, which means that "latest dev build" is sometimes actually a PR.

**Comment by QarthO** [comment]
GP does have an existing actions workflow: 
https://github.com/GriefPrevention/GriefPrevention/blob/master/.github/workflows/ci.yml

Looks like it's never called or ran tho

**Comment by RoboMWM** [comment]
Feature: github action for uploading alpha/beta build to bukkitdev/modrinth and possibly github releases as pre-release.

So originally I was interested in automated uploads to bukkitdev, and I guess now modrinth since I suppose they have an API too. Would be good for beta/alpha release channels. That I'd be interested in a github action for, tho I'd prolly wanna trigger that manually only, at least for right now. That way users don't have to rely on CI builds. I guess it could also do pre-releases.

**Reply by Jikoo** [reply]
https://github.com/Kir-Antipov/mc-publish
Haven't used this one myself yet, but am looking to swap because the one I currently use is running Node 16 and will stop working soon.

**Comment by RoboMWM** [comment]
Security concern: Restrict github actions token from being able to modify/delete anything in github (e.g. prohibit branch force-push, deleting/modifying prior github releases, etc.)

I don't want Github actions to be able to modify anything already existing on github. Is ok if it creates new stuff, but primary concern is deletion/modification of already-existing things, be it commits, releases, issues, comments, etc. Want to make these entirely off-limits and no privilege escalation possible within its own configuration.

> To clarify, I guess I should not label this as a security concern per-se, but instead something to avoid mistakes. If I (or another person) makes a change to a workflow yml, and it happens to not behave as expected, does it have the possibility to perform these operations? E.g., I know if I make changes to the appveyor yml, all the chaos would be contained in appveyor only.

**Reply by RoboMWM** [reply]
To clarify, I guess I should not label this as a security concern per-se, but instead something to avoid mistakes. If I (or another person) makes a change to a workflow yml, and it happens to not behave as expected, does it have the possibility to perform these operations? E.g., I know if I make changes to the appveyor yml, all the chaos would be contained in appveyor only.

**Reply by nouish** [reply]
It will not write to or otherwise change the repository unless you set up an action with the necessary permissions to do so.

**Reply by nouish** [reply]
See https://github.com/nouish/GriefPrevention/blob/main/.github/workflows/ci.yml for a typical action setup with dependency caching and artifacts.

**Comment by RoboMWM** [comment]
Question/security: How would github ci test PRs?

I'm guessing from what I've glanced at before, it uses a token with the PR's privileges and/or makes it read-only? Can't recall how this functions.

**Reply by nouish** [reply]
Again, you're overthinking this.

If this was an active problem would-be hackers would be targeting the likes of apache/maven repo and what not.

GitHub Actions use the configurations in your primary branch **ONLY**.

That is to say, if I were to open a PR with changes to `.github/workflows/ci.yml` it would still use the configuration already on the `master` branch.

**Reply by nouish** [reply]
Add `permissions: {}` and be done with the paranoia.

**Reply by RoboMWM** [reply]
This is a question strictly on PRs, thank you for the info. I'm not too concerned about this one (as I have labeled it as question primarily), just wanted clarification on this aspect cuz I'm too lazy rn to re-read the documentation myself for this atm and I put it here as part of a list of other things I want to know about actions.

**Comment by RoboMWM** [comment]
Here are my current thoughts on this proposal: The only thing mentioned thus far in favor of using github actions is a preference for its UI. 

My concerns continue to be permissions to github repo settings can be granted by simply modifying the workflow yml, which means messing around with the workflow yml is a bit more dangerous than messing around with appveyor.yml. (Yes, I know it's a matter of not being stupid when editing the files and not blindly pulling PRs, and it's unlikely, but the possibility still exists, and I do have to change security settings to allow for these workflows to run.)

If github actions is significantly faster than appveyor in performing PR status checks, then I'll be more apt to consider it... Otherwise, I don't really see the benefit of me using a different Actions UI. Yes, selfish decision, whatever, but I don't see any benefit functionally; it seems to me that this proposal's motivation is just an aesthetic choice. I'm not going to flip a security switch for that (unless the current solution is more vulnerable, whether to legitimate security issues or mistakes... which I don't think it is. Appveyor can't obtain any access to GitHub.)

(Yes I'm being repetitive, but it seems so are the responses I get to my concerns. I don't see how my concerns are invalid. I get it that others are doing this, and my guess is these projects either do more with GitHub Actions, started out this way, or just want to have everything together in one service. A service offering provided by GitHub directly is gonna get more visibility than seeking out a 3rd party solution. Just because a lot of people use something doesn't mean it's secure. Just that it's probably good enough for use... and imma be honest, this is annoying to see as a 'counter' to my concerns. Functionally I don't see the difference.)

That being said - if there are ways to backup issues and discussions, I probably will enable Actions soon after implementing such. It wouldn't change the concerns I have though, but at least I'd be a little bit less worried in the rare event something does modify/delete existing content in this repo/issues/discussions.

And yes, I know there are workflows that can run in conjunction with GitHub rules to block PRs that edit workflow yamls, but it doesn't prevent admins (i.e. me) from experimenting and making dangerous mistakes.

To summarize:
GitHub actions:
- 👍Pleasing UI (subjective)
- 👍Integrated directly on GitHub website
- 👎Workflow yml can be edited to elevate its permissions to modify GitHub repo, issues, discussions

Appveyor:
- 👍Cannot edit GitHub repo, issues, discussions, etc. (without explicitly being given a GitHub token of a repo administrator)
- 👎Collects and spits out builds of pull requests without explicitly labeling as such

Please let me know if there's something I'm missing, incorrect about, or not understanding here.

**Reply by RoboMWM** [reply]
Some followup on this from #robomwm
> [15:13:37] Dumcord: <g​ranny.dev> idm if the PRs don't end up getting merged - if anything, they can serve as a reference if you choose to use github actions in the future
[15:14:06] @RoboMWM: sounds good, thank you! Ye it do be a reference
[15:15:21] @RoboMWM: I'd really like to use it especially if ppl want it, but the things I gotta do to allow it concern me a bit and seem to have no answer other than "likely not gonna happen" sorta deal which is not how I wanna deal with things I care about lol
[15:15:46] Dumcord: <g​ranny.dev> i can understand that \:)
[15:16:08] @RoboMWM: that and I see no extra benefit other than another place for dev builds of GP (without PR builds too ig)
[15:16:56] @RoboMWM: If it provided something else that appveyor doesn't already do, I'd be more inclined ¯\_(ツ)_/¯
[15:17:27] Dumcord: <g​ranny.dev> there's a system that can be added that's similar to Paper's where if you add a label to the PR, it creates a build, then adds a link to said build in the PR description, ex. https://github.com/PaperMC/Paper/pull/ 10231
[15:17:45] Dumcord: <g​ranny.dev> i say similar - i pretty much copied and slightly modified it for use with Pl3xMap lol
[15:22:40] RoboMWM: yea my approach with more serious projects is to avoid giving automated access to edit other people's stuff or allow for it to be potentially added, so very unlikely I'll add something like that lol. Or if I ever do, at least to have it either post it as a comment from itself instead of editing description. I never added dependabot back in the day cuz it could be configured to auto-accept PRs. They've removed that now, I think in favor of Actions? But ya.
[15:23:57] RoboMWM: But if I can get backups of this stuff, and better yet something that could monitor and diff such changes (a repo of issues/discussions would be nice lol), that would alleviate most of my concerns.

**Reply by QarthO** [reply]
Something to note about GitHub Actions is you only get 2000min/month of free runtime per organization. So this will include all the other CI scripts you have for the other repos such as the docs and website. Realistically we won't ever hit that threshold even if we're running several builds every day, on each repo. But the concern would be valid if we had PRs building artifacts, that someone could spam PRs and run up our actions minutes. However, GitHub by default requires first-time contributors to have their workflows approved (check [here](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks#about-workflow-runs-from-public-forks))

If we continue with Appveyor, I'm also pretty sure you can modify the Appveyor config so that it'll properly label PRs, (briefly looking at it, it does seem a lot more complicated than GitHubs method)

I'm all for GAs, I think being directly integrated with GitHub is a huge benefit in terms of usability and accessibility, (not just a Pleasing UI, which I do agree with you, that it should be a non-factor in the decision to switch)

**Reply by QarthO** [reply]
>  avoid giving automated access to edit other people's stuff


Also I don't see how this is an automated process to modify someones elses stuff. Its not all. Someone has to manually modify the workflow.

**Reply by Jikoo** [reply]
I feel like that's a relatively good summary. It is missing the fact that Actions can automate workflows you've expressed issues with in the past, namely cutting releases. I don't touch jars for OpenInv, those get compiled and added to a draft release by Actions when I push a new tag. Then, when I'm happy and press the green button, the jar is sent over to CurseForge with a copy of the release details.

I get where you're coming from re: security, I just don't think you'd make a mistake of that scale tbh.
Re: accidentally muddling something severe: Worst case scenario you might run an extra build or create an incorrect post if a conditional evaluated wrong. In general, imagine what would happen if the conditional for the thing you want to do evaluated incorrectly - that's the worst the nasty side effect could be. Since you'd never support deleting/merging automatically and would as a result never include code that could do that, it wouldn't be a risk of a bad conditional. The last time I was worried about actually messing up with Actions, I made a [test repo](https://github.com/Jikoo/ActionsTest) and figured out the problems there, then modified the functional files to do what I actually wanted them to.

Actions does generally outperform Appveyor (and I actually had problems where AppVeyor would skip builds when done at high load times like 5AM on the first of a month when GitHub used to do all the monthly Dependabot checks) but not by a wide enough margin that it really makes any difference. May be more of a concern as we add more tests.

I actually think AppVeyor has a much nicer UI overall for a CI (why does everything on GH have these massive spaces that look like a mobile UI?), just the fact that it can't be filtered to say "yes, user who is going to do something dangerous, here are official development builds from the master branch and not some random PR" is pretty rough.

**Reply by RoboMWM** [reply]
Yea, I think the original reason why I added appveyor was to have a convenient way to download and try test builds. Not necessarily for others. But it is what it is, after all, I often did (and still do for some) go to the CI for plugins I've used on my server before I was a plugin developer.

> It is missing the fact that Actions can automate workflows you've expressed issues with in the past, namely cutting releases.

Yea, and then GitHub at some point added a button to do exactly that lol. Gonna be honest, I've been relying on that. And now with AI being a thing... maybe in the future... lol. But I don't cut releases often enough anymore to warrant needing to do that. I think it was a larger concern back when I was maintaining multiple plugins at the time, so it seemed like a laborious chore to publish release notes everywhere. Now that I'm focusing on just GP, that need/want has diminished significantly.

Thank you for the feedback though! Maybe at some point GitHub will provide a way to restrict permissions of an action no matter what the workflow yml specifies, maybe; apparently it started out that they were granted a permissive token by default, and only later did did they add the default permission settings.

**Comment by QarthO** [comment]
I started this discussion, but I would like to add that I believe the differences between the ci's and in our current use cases are pretty much negligible. I believe that switching to Actions will __**not**__ make any major improvements. 

An analogy that I think seems fit is like the decision from upgrading to a new phone, from a new-ish phone. Lets say you have the iPhone 13, (Appveyor) and the new iPhone 14 (Actions) comes out. Is it nicer? Definitely. Does it have more features that will be integrated with the new iOS version that also came out (github/the repo), yes 100%. Do you need to upgrade to the new phone? No, not at all. Your new~ish phone is still great as is.

The only difference I see is that it doesn't cost us an arm and a leg ($1000 to get the new phone) to switch to Actions.
